homeassistant/components/nws/weather.py

```diff
@@ -165,85 +73,71 @@ def convert_condition(time, weather):
-    # ID request as being from HA, pynws prepends the api_key in addition
-    api_key_ha = f"{api_key} homeassistant"
-    nws = SimpleNWS(latitude, longitude, api_key_ha, websession)
+    nws_data = hass.data[DOMAIN][base_unique_id(latitude, longitude)]
 
-    _LOGGER.debug("Setting up station: %s", station)
     try:
-        await nws.set_station(station)
-    except ERRORS as status:
-        _LOGGER.error(
-            "Error getting station list for %s: %s", (latitude, longitude), status
-        )
+        await nws_data.async_set_station(station)
+    except (aiohttp.ClientError, asyncio.TimeoutError) as err:
+        _LOGGER.error("Error automatically setting station: %s", str(err))
         raise PlatformNotReady
 
-    _LOGGER.debug("Station list: %s", nws.stations)
-    _LOGGER.debug(
-        "Initialized for coordinates %s, %s -> station %s",
-        latitude,
-        longitude,
-        nws.station,
-    )
+    await nws_data.async_update()
 
-    async_add_entities([NWSWeather(nws, mode, hass.config.units, config)], True)
+    async_add_entities(
+        [
+            NWSWeather(nws_data, "daynight", hass.config.units),
+            NWSWeather(nws_data, "hourly", hass.config.units),
+        ],
+        False,
+    )
 
 
 class NWSWeather(WeatherEntity):
     """Representation of a weather condition."""
 
-    def __init__(self, nws, mode, units, config):
+    def __init__(self, nws, mode, units):
         """Initialise the platform with a data instance and station name."""
         self.nws = nws
-        self.station_name = config.get(CONF_NAME, self.nws.station)
+        self.station = nws.station
+        self.latitude = nws.latitude
+        self.longitude = nws.longitude
+
         self.is_metric = units.is_metric
         self.mode = mode
 
         self.observation = None
         self._forecast = None
 
-    @Throttle(MIN_TIME_BETWEEN_UPDATES)
-    async def async_update(self):
-        """Update Condition."""
-        _LOGGER.debug("Updating station observations %s", self.nws.station)
-        try:
-            await self.nws.update_observation()
-        except ERRORS as status:
-            _LOGGER.error(
-                "Error updating observation from station %s: %s",
-                self.nws.station,
-                status,
-            )
-        else:
-            self.observation = self.nws.observation
-        _LOGGER.debug("Observation: %s", self.observation)
-        _LOGGER.debug("Updating forecast")
-        try:
-            if self.mode == "daynight":
-                await self.nws.update_forecast()
-            else:
-                await self.nws.update_forecast_hourly()
-        except ERRORS as status:
-            _LOGGER.error(
-                "Error updating forecast from station %s: %s", self.nws.station, status
-            )
-            return
+    async def async_added_to_hass(self) -> None:
+        """Set up a listener and load data."""
+        async_dispatcher_connect(
```

Comment:
Please undo the dispatcher in 
```async def async_will_remove_from_hass(self):```

```suggestion
        self._dispatcher_undo = async_dispatcher_connect(
```

Reply:
Or better wrap with `self.async_on_remove(`